xrpl/transaction/amm_deposit.go

```diff
@@ -1,14 +1,181 @@
+	tfSingleAsset uint = 524288
+	// Perform a double-asset deposit with specified amounts of both assets.
+	tfTwoAsset uint = 1048576
+	// Perform a single-asset deposit and receive the specified amount of LP Tokens.
+	tfOneAssetLPToken uint = 2097152
+	// Perform a single-asset deposit with a specified effective price.
+	tfLimitLPToken uint = 4194304
+	// Perform a special double-asset deposit to an AMM with an empty pool.
+	tfTwoAssetIfEmpty uint = 8388608
+)
+
+// Perform a double-asset deposit and receive the specified amount of LP Tokens.
+func (a *AMMDeposit) SetLPTokentFlag() {
+	a.Flags |= tfLPToken
+}
+
+// Perform a single-asset deposit with a specified amount of the asset to deposit.
+func (a *AMMDeposit) SetSingleAssetFlag() {
+	a.Flags |= tfSingleAsset
+}
+
+// Perform a double-asset deposit with specified amounts of both assets.
+func (a *AMMDeposit) SetTwoAssetFlag() {
+	a.Flags |= tfTwoAsset
+}
+
+// Perform a single-asset deposit and receive the specified amount of LP Tokens.
+func (a *AMMDeposit) SetOneAssetLPTokenFlag() {
+	a.Flags |= tfOneAssetLPToken
+}
+
+// Perform a single-asset deposit with a specified effective price.
+func (a *AMMDeposit) SetLimitLPTokenFlag() {
+	a.Flags |= tfLimitLPToken
+}
+
+// Perform a special double-asset deposit to an AMM with an empty pool.
+func (a *AMMDeposit) SetTwoAssetIfEmptyFlag() {
+	a.Flags |= tfTwoAssetIfEmpty
 }
 
 func (*AMMDeposit) TxType() TxType {
 	return AMMDepositTx
 }
 
-// TODO: Implement flatten
-func (s *AMMDeposit) Flatten() FlatTransaction {
-	return nil
+func (a *AMMDeposit) Flatten() FlatTransaction {
+
+	// Add BaseTx fields
+	flattened := a.BaseTx.Flatten()
+
+	// Add AMMDeposit-specific fields
+	flattened["TransactionType"] = "AMMDeposit"
+
+	flattened["Asset"] = a.Asset.Flatten()
+	flattened["Asset2"] = a.Asset2.Flatten()
+
+	if a.Amount != nil {
+		flattened["Amount"] = a.Amount.Flatten()
+	}
+
+	if a.Amount2 != nil {
+		flattened["Amount2"] = a.Amount2.Flatten()
+	}
+
+	if a.EPrice != nil {
+		flattened["EPrice"] = a.EPrice.Flatten()
+	}
+
+	if a.LPTokenOut != nil {
+		flattened["LPTokenOut"] = a.LPTokenOut.Flatten()
+	}
+
+	if a.TradingFee != 0 {
+		flattened["TradingFee"] = a.TradingFee
+	}
+
+	return flattened
+}
+
+func (a *AMMDeposit) Validate() (bool, error) {
+	_, err := a.BaseTx.Validate()
+	if err != nil {
+		return false, err
+	}
+
+	if ok, err := IsAsset(a.Asset); !ok {
+		return false, err
+	}
+
+	if ok, err := IsAsset(a.Asset2); !ok {
+		return false, err
+	}
+
+	if a.Amount2 != nil && a.Amount == nil {
+		return false, errors.New("ammDeposit: must set Amount with Amount2")
+	} else if a.EPrice != nil && a.Amount == nil {
+		return false, errors.New("ammDeposit: must set Amount with EPrice")
+	} else if a.LPTokenOut == nil && a.Amount == nil {
+		return false, errors.New("ammDeposit:  must set at least LPTokenOut or Amount")
+	}
+
+	if ok, err := IsAmount(IsAmountArgs{field: a.Amount, fieldName: "Amount", isFieldRequired: false}); !ok {
```

Comment:
We should remove the `IsAmountArgs` struct and spread the props. It is much easier to read

Reply:
I've done the update in #43 , that way I won't have merge conflicts later.